grails-app/domain/org/pih/warehouse/shipping/Shipment.groovy

```diff
@@ -697,5 +697,63 @@ class Shipment implements Comparable, Serializable {
     Boolean hasChildContainer() {
         shipmentItems?.any { it?.container }
     }
+
+    Set<Event> getCustomEvents() {
+        return events.findAll { it.eventType?.eventCode?.name()?.startsWith("CUSTOMS") }
+    }
+
+    @Override
+    ReferenceDocument getReferenceDocument() {
+        return new ReferenceDocument(
+                label: shipmentNumber,
+                url: "/stockMovement/show/${requisition?.id ?: id}",
+                id: id,
+                identifier: shipmentNumber,
+                description: description,
+                name: name,
+        )
+    }
+
+    @Override
+    List<HistoryItem> getHistory() {
+        List<HistoryItem> histories = []
+        ReferenceDocument shipmentReferenceDocument = getReferenceDocument()
+        // First collect history of CREATED event
+        histories.add(new HistoryItem<Shipment>(
+                identifier: shipmentNumber,
+                date: dateCreated,
+                location: origin,
+                eventCode: EventCode.CREATED,
+                referenceDocument: shipmentReferenceDocument
+        ))
+        // Then collect history of a shipped event if any
+        if (hasShipped()) {
+            histories.add(new HistoryItem<Shipment>(
+                    identifier: shipmentNumber,
+                    date: dateShipped(),
+                    location: origin,
+                    eventCode: EventCode.SHIPPED,
+                    referenceDocument: shipmentReferenceDocument,
+            ))
+        }
+        // Then collect history of a received event if any
+        if (wasReceived()) {
```

Comment:
Should not be partially received before this one?

Reply:
Yes, but it shouldn't matter since we *should* sort before displaying. With that said, i think dding methods to HistoryItem to ensure the correct ordering would be a good idea as well.

Reply:
It depends on if we want to keep the ordering in this method already. I assumed this could be sorted in every possible way on the view side, but I can change the ordering.

Reply:
I switched the order